app/src/main/java/org/onionshare/android/tor/TorManager.kt

```diff
@@ -231,4 +295,100 @@ class TorManager @Inject constructor(
         return File(dataDir, "ControlSocket").absolutePath
     }
 
+    private suspend fun checkStartupProgress() {
+        LOG.info("Starting check job")
+
+        var usedBuiltInBridges = false
+        runWhenMoatTimeIsUp {
+            val bridges = try {
+                getBridgesFromMoat()
+            } catch (e: IOException) {
+                LOG.warn("Error getting bridges from moat: ", e)
+                null
+            }
+            if (bridges.isNullOrEmpty()) {
+                // fallback to built-in bridges if moat doesn't return anything
+                useBridges(meekBridges + snowflakeBridges + obfs4Bridges)
+                usedBuiltInBridges = true
+            } else {
+                useBridges(bridges)
+            }
+        }
+        // give moat some more time to succeed
+        delay(MOAT_TIME_SINCE_START)
+        // moat can give us slow bridges, so try harder
+        delay(MOAT_TIME_SINCE_START)
+        if (usedBuiltInBridges) {
+            // let's try with without bridges again, just in case
+            val conf = listOf(
+                "UseBridges 0",
+            )
+            controlConnection?.setConf(conf)
+        } else {
+            // try built-in bridges if we haven't already
+            useBridges(meekBridges + snowflakeBridges + obfs4Bridges)
```

Comment:
I wonder if we could simplify things here. If Moat returns some bridges then we're trying those within the block above, and then we have to try the defaults here. Whereas if Moat doesn't return any bridges then we're trying the default bridges in the block above, and then we don't need to try them again here. Either way, we finally try again without bridges. Feels like it should be possible to simplify that to a single chain:
* Try without bridges
* Try Moat bridges, if any
* Try default bridges
* Try without bridges again

If this is what we're doing in effect then why does the flow control need to be so complicated?